### PR TITLE
BUG: fix and lint for undefined names F821

### PR DESCRIPTION
**Pull request by jbrockmendel**
This is one of the most important ones.  A serious bug that is easy to prevent and hard to debug.

See comments inline.

### CODE REVIEW

File: statsmodels/miscmodels/try_mlecov.py [hunk 1]
```diff
@@ -112,18 +114,9 @@ def mvn_nloglike_obs(x, sigma):
     return llike
 
 def invertibleroots(ma):
-    import numpy.polynomial as poly
-    pr = poly.polyroots(ma)
-    insideroots = np.abs(pr)<1
-    if insideroots.any():
-        pr[np.abs(pr)<1] = 1./pr[np.abs(pr)<1]
-        pnew = poly.Polynomial.fromroots(pr)
-        mainv = pn.coef/pnew.coef[0]
```
Comment (jbrockmendel) [inline statsmodels/miscmodels/try_mlecov.py]: `pn` is undefined.  Could reasonably be typo for either `pr` or `pnew`.  Turns out to be `pnew`.  Also turns out we have another strictly-dominating option that does exactly the same thing.

File: statsmodels/sandbox/distributions/gof_new.py [hunk 2]
```diff
@@ -526,8 +526,8 @@ def asquare(cdfvals, axis=0):
     islice = [None] * ndim
     islice[axis] = slice(None)
     slice_reverse[axis] = slice(None, None, -1)
-    asqu = -((2. * np.arange(1., nobs+1)[islice] - 1) *
-            (np.log(cdfvals) + np.log(1-cdfvals[slice_reverse]))/nobs).sum(axis) \
```
Comment (jbrockmendel) [inline statsmodels/sandbox/distributions/gof_new.py]: This is needed to fix a FutureWarning produced by numpy when actually running the code below.

File: statsmodels/sandbox/distributions/gof_new.py [hunk 3]
```diff
@@ -698,7 +698,7 @@ def rvs(self, args, size):
     tmp = [bootstrap(NewNorm(), args=(0,1), nobs=nobs, nrep=1,
                      value=0.576/ (1 + 4./nobs - 25./nobs**2),
                      batch_size=10) for _ in range(10)]
-    resu3 = np.array(resu).mean()
+    resu3 = np.array(tmp).mean()
```
Comment (jbrockmendel) [inline statsmodels/sandbox/distributions/gof_new.py]: Using `tmp` here, the `assert_array_almost_equal` below succeeds.

File: statsmodels/sandbox/stats/multicomp.py [hunk 4]
```diff
@@ -1424,14 +1424,17 @@ def distance_st_range(mean_all, nobs_all, var_all, df=None, triu=False):
...
+    std_pairs_ = np.sqrt(var_pairs)
 
     idx1, idx2 = np.triu_indices(n_means, 1)
     if triu:
         #select all pairs from upper triangle of matrix
         meandiffs = meandiffs_[idx1, idx2]
         std_pairs = std_pairs_[idx1, idx2]
+    else:
+        meandiffs = meandiffs_
+        std_pairs = std_pairs_
```
Comment (jbrockmendel) [inline statsmodels/sandbox/stats/multicomp.py]: This is a best-guess at what is intended.  The code here is very similar to the `tukeyhsd` function.

Comment (bashtage) [reply statsmodels/sandbox/stats/multicomp.py]: Looks right. Perhaps a comment 

\# These were undefined, and this appears to be the most likely intent, verify before moving out of sandbox

**Comment by ChadFulton** [comment]
It was my understanding that two maintainers (@josef-pkt and me) were opposed to you linting sandbox files. Unless @josef-pkt has changed his mind, please stop submitting PRs with changes of any kind to sandbox.

**Comment by bashtage** [comment]
This PR is not linting, but is mostly bug fixing.  The code that is modified is wrong, in the sense that it cannot run.  In all instances except one the fix is clearly correct (in this one, he has to guess the author's intent, which is why I suggested a comment).

**Comment by josef-pkt** [comment]
> This PR is not linting, but is mostly bug fixing.

Some are good, but overall this is not bug-fixing, this is making flake happy.

Bug fixing means you get a correctly working function with unit tests as outcome !!!
Some "fixes" here change code in non-working, draft functions that won't work after the flake fix either.
Or even if they fix things than it is based on guessing and not on unit tests

**Comment by josef-pkt** [comment]
As I told @jbrockmendel before
It's worse to have a function returning wrong numbers, than having a function that raises because of a bug.

**Comment by bashtage** [comment]
I don't understand the standard.  There is tons of code that has been accepted that is wrong (as in doesn't even run) and obviously completely untested.  This type of patch is progress, and it is unreasonable to me to start expecting perfection when this isn't close to the standard used throughout. Perfect is the opposite of good.  It is not as-if this code is being moved out of its current location or users are being encouraged to use it.

**Comment by josef-pkt** [comment]
i.e. the PR can be rejected because it is making substantive changes without accompanying unit tests.

**Comment by ChadFulton** [comment]
I don't understand the concern that prompts making changes to sandbox code, since it is clearly labeled as "sandbox" and it is not part of the API. @josef-pkt prefers for it to remain unmodified, and I don't see any benefit on the other side (the code still probably doesn't run all the way through, and it isn't going to be called by another). To me, this is enough to just forget that the sandbox folder even exists (which is what I do :) ).

**Comment by bashtage** [comment]
The gain is that it helps improve code standards by simplifying enforcement of a set fo widely adopted conventions on new code.  Not that it is especially helpful for sandbox (even though it is here IMO). 

I think sandbox should not be under `statsmodels\` and instead should be in git but under root, like `examples\`.  This way it is clear that it isn't part of any shipped code, and can't be relied upon.  If someone wants to promote sandbox to production, then move it to a module and add the tests.  Just by $.02.

**Comment by josef-pkt** [comment]
>  on new code

you can just exempt the sandbox from any pep-8/flake enforcement, so Mr. J Brock stops changing things all over the place.

Note, this PR has some fixes that are most likely good, but based on guessing the intention and not unit tests.

**Comment by jbrockmendel** [comment]
> I don't understand the concern that prompts making changes to sandbox code, since it is clearly labeled as "sandbox" and it is not part of the API

@ChadFulton Code is still imported from outside of sandbox (which btw you're on record as being against in #4887), so I have to trust it, or at least keep track of which parts (#4423 "I have no overview") are actually used/tested/documented/...  My attempts to get the ready-for-primetime parts out of sandbox  (most recently #5745) have been refused.  These moves would be necessary before we could implement @bashtage's suggestion above to move statsmodels/sandbox/ to just sandbox/.

> since it is clearly labeled as "sandbox"

It is labeled as sandbox, and the `sandbox.__init__` file helpfully tells us "This is sandbox code", but that doesn't explain what it actually means.  I've definitely seen GH Issues opened because some user tried to use `sandbox.tsa.garch` (see also #5718).

As long as sandbox is used/available from outside of sandbox, I have to treat it like I would any other code.

**Comment by jbrockmendel** [comment]
> It's worse to have a function returning wrong numbers, than having a function that raises because of a bug.

In such a case, the bug that is raised should be `NotImplementedError`.

**Comment by josef-pkt** [comment]
> In such a case, the bug that is raised should be NotImplementedError.

In sandbox functions we want to get an exception that provides a starting point for proper bugfixes or for finishing up the function.

Outside the sandbox bugs should be properly fixed, including unit tests.
NotImplementedErrors are for cases where an enhancement is missing and are not a replacement for bugfixes.

File: statsmodels/sandbox/stats/multicomp.py (hunk 4 above)
Comment (jbrockmendel) [reply statsmodels/sandbox/stats/multicomp.py]: It looks like this problem was introduced in ec116cce7619219ce9898732aab19ca2881e6e0a (no associated PR AFAICT) in Jan 2011.  Previously the L1430-L1434 block was commented-out.

**Comment by bashtage** [comment]
F821 passes and is in the standard lint run.

**Pull request closed without merging**